### PR TITLE
Prototype i32x4.widen_i8x16_{s,u}

### PR DESCRIPTION
**Pull request by tlively**
As proposed in https://github.com/WebAssembly/simd/pull/395. Note that the other
instructions in the proposal have not been implemented in LLVM or in V8, so
there is no need to implement them in Binaryen right now either. This PR
introduces a new expression class for the new instructions because they uniquely
take an immediate argument identifying which portion of the input vector to
widen.

### CODE REVIEW

**Comment by tlively** [comment]
cc @omnisnip.

@kripken, I enabled auto-merge on the PR, so make sure to disable first it if you want to LGTM+nits.

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -5466,6 +5469,27 @@ bool WasmBinaryBuilder::maybeVisitSIMDLoadStoreLane(Expression*& out,
...
+    default:
+      return false;
+  }
+  auto* curr = allocator.alloc<SIMDWiden>();
+  curr->op = op;
+  curr->index = getLaneIndex(4);
+  curr->vec = popNonVoidExpression();
+  curr->finalize();
+  out = curr;
+  return true;
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp:5490]: please use a Builder and call the method on there.

File: src/wasm/wasm-s-parser.cpp [hunk 2]
```diff
@@ -1666,6 +1666,15 @@ SExpressionWasmBuilder::makeSIMDLoadStoreLane(Element& s,
...
 }
 
+Expression* SExpressionWasmBuilder::makeSIMDWiden(Element& s, SIMDWidenOp op) {
+  auto* ret = allocator.alloc<SIMDWiden>();
+  ret->op = op;
+  ret->index = parseLaneIndex(s[1], 4);
+  ret->vec = parseExpression(s[2]);
+  ret->finalize();
+  return ret;
+}
```
Comment (kripken) [inline src/wasm/wasm-s-parser.cpp:1676]: ditto

**Review by kripken — APPROVED** [review]
lgtm % comments

**Comment by kripken** [comment]
Heh, I did not read the later comment and so it did an auto-merge... anyhow, my comments were not urgent.

**Comment by omnisip** [comment]
What do I need to do to test this with emscripten and clang?

**Comment by kripken** [comment]
In a few hours there should be a new tip of tree build with this included. You can then do `emsdk install tot` to get it.

Or, you can build binaryen manually and point emscripten to that location by editing the `.emscripten` config file.

**Comment by omnisip** [comment]
Okay cool! Will let you guys know how it goes.

**Pull request merged**